### ISSUE TITLE
`untar` is missing seed support

### ISSUE THREAD

**Issue by RyuzakiKK** [folbricht/desync#242]
In our use case, we have an A/B setup where we apply block based system updates with `desync extract` and use the active partition as a seed.
For very small updates, using a block based `.caibx` introduces a lot of overhead. So we were evaluating to produce file based `.caidx` updates in those circumstances, as an alternative to speed up the updates when just a handful of files change.

However I noticed that `desync untar` doesn't have any seed support at the moment.

How a seed for `untar` should be designed? If we produce a `.caidx` of a local directory, can we grab it's corresponding chunk ID without having to tar the entire local directory first?

I can see that Casync handles `.caidx` with `extract`, and if you provide a path as a seed, it starts its process by seeding one by one all the files in the provided directory.

**Comment by folbricht** [comment]
I'm not sure it's possible or useful to have seeds for untar. Are you thinking about seeds to restore the full archive on the target and then untar? Or the untar stage itself? The former is possible already if you split up the process into the extract and untar stages. The latter may not be possible.

Wrt the untar stage, the issue is that with seeds you can check if a chunk is present or not. But there aren't any chunks available if the target is a directory. A caidx is just a caibx of an archive (catar), and there's no concept of chunks of the target. A chunk inside an archive can span over multiple files so you couldn't really say a file is changed or not until it's unpacked. If the file inside the archive is large, there's a chance that chunking the target files individually first may give you some overlap of chaunkIDs, but it's not going to work for smaller files.

If you have the disk space, can you maintain the caidx+catar file on the target with make/extract, then use untar (without -i) to restore the directory? This will let you use a seed. Or are you trying to save write I/O?

**Comment by RyuzakiKK** [comment]
Our primary goals are to reduce the download size and speed up the whole update process. Reducing I/O is nice to have, but not a primary concern at the moment.

If we produce two system images where just a few files changed, a `.caibx` will still require to download a lot of chunks from the store because, for example, the metadata of all the files changed too.

With `.caidx` (and `--no-time`) I noticed that instead we were able to significantly reduce the number of required chunks, especially for small system updates.
Whether that translates to faster updates or not is still unclear. I'll run some benchmark to confirm.

> Wrt the untar stage, the issue is that with seeds you can check if a chunk is present or not. But there aren't any chunks available if the target is a directory. A caidx is just a caibx of an archive (catar), and there's no concept of chunks of the target. A chunk inside an archive can span over multiple files so you couldn't really say a file is changed or not until it's unpacked. If the file inside the archive is large, there's a chance that chunking the target files individually first may give you some overlap of chaunkIDs, but it's not going to work for smaller files.
>
> If you have the disk space, can you maintain the caidx+catar file on the target with make/extract, then use untar (without -i) to restore the directory? This will let you use a seed. Or are you trying to save write I/O?

Is this basically what Casync is doing? But instead of writing the catar file on disk they keep it in memory?
I can check Casync implementation to see exactly how they are handling the seed part there.

I'll also try to create a catar as a first step and benchmark how long that takes. If creating a catar and extracting a caidx doesn't bring any real speed advantage, that might not be worth for us. In that case we'll probably just stick with caibx and evaluate how to further reduce images differences.

**Comment by folbricht** [comment]
Sounds like what you need is:
1. Tar the files up on the source system with `desync tar --no-time`. You could use -i or use a two-step process with tar + make (faster)
2. On the target, you you build the catar file with`desync extract ... file.caidx` - The transfer should be quick since you can use seeds or caches
3. On the target you extract the archive with `desync untar` (no -i)

**Comment by RyuzakiKK** [comment]
Given that on the target we just have the extracted archive files around, unless I'm missing something, the actual steps will end up being:

On the source:
1. `desync tar --no-time`

On the target:
2. `desync tar --no-time`,  to create a usable seed from the files in the currently active partition
3. `desync extract`
4. `desync untar`
5. Delete the generated seed catar and the downloaded catar update

And in order to successfully apply that we would need a free space that is at least 2x the full update size. That's a substantial tradeoff compared to the `caibx` where we don't need any reserved free space at all.

To skip the step 2. we would need to keep the seed catar always around, but always using all that disk space is probably unfeasible.

The more I look and this, the more I suspect `caidx` will sadly not help us to speed up small system updates.

**Comment by folbricht** [comment]
Correct, if you can't keep the catar on disk on the target it's going to require some extra time, either for processing, or for transferring.

Something that's not yet implemented which could help here is to introduce something like a "DirectorySeed" where desync scans a directory and chunks all  large files into a kind of index without storing the chunks. The seed index would contain a map of chunkid => {filepath, position in the file}. Then the untar would be able to use chunks directly out of target files. Building this Seed is going to require processing the target files again so it'll add some time. Though that could be faster than transferring the chunks, depending on the situation.

I suppose this index could even be persisted to disk for the next update.

**Comment by folbricht** [comment]
Perhaps this seed index could be produced as part of the untar operation, to then help find chunks during future untar.

**Comment by RyuzakiKK** [comment]
Yes, that could definitely help.

If we decide to actually use `untar` for small updates, I can probably try to implement something like that.

**Comment by RyuzakiKK** [comment]
> Given that on the target we just have the extracted archive files around, unless I'm missing something, the actual steps will end up being:
> 
> On the source:
> 1. `desync tar --no-time`
> 
> On the target:
> 2. `desync tar --no-time`, to create a usable seed from the files in the currently active partition
> 3. `desync extract`
> 4. `desync untar`
> 5. Delete the generated seed catar and the downloaded catar update
> 
> And in order to successfully apply that we would need a free space that is at least 2x the full update size. That's a substantial tradeoff compared to the `caibx` where we don't need any reserved free space at all.

I guess that another alternative here would be to merge `desync extract` and `desync untar` into a single command.
In this way we could unpack the catar on the fly without having to write it on the disk first.

More or less the same concept of `desync untar -i`, which starts to extract the catar as it assembles it.

**Comment by lights0123** [comment]
I'm interested in implementing this for our own IoT updating system. Any tips on where to start? My ideal workflow would be:

- Create new updated image.
- Create index of this image and save it within (the proposed "DirectorySeed" mechanism)
- `desync tar` this new image to create caidx and cacnk
- `desync untar` these update files on the target using the index already on-disk as a seed
- Mark this update read-only so the index on-disk will always be valid

Instead of using A/B partitions, each version is a btrfs subvolume so that updates can reflink the majority of shared files—keeping many past versions is important for our case, and this allows that without extreme disk usage.

**Comment by folbricht** [comment]
What does "image" in this refer to? Normally I think of an image as a single file, but then `tar`/`untar` wouldn't be needed so perhaps in this case "image" is a directory of multiple files?

When dealing with `untar` and seeds, we'd first need to define what a seed for a directory actually looks like. Perhaps another directory of similar files? But during untar while reading the stream, how would you know that the file being unpacked matches one already on disk. You'd have to fully unpack it first and then compare which doesn't really help with being fast or efficient.

**Comment by lights0123** [comment]
> so perhaps in this case "image" is a directory of multiple files

Yeah, I didn't word that the best. It's a Linux root filesystem that is packaged as a GPT disk image for initial factory deployment, but I'd like future updates to happen at the filesystem level.

> When dealing with untar and seeds, we'd first need to define what a seed for a directory actually looks like. Perhaps another directory of similar files? But during untar while reading the stream, how would you know that the file being unpacked matches one already on disk. You'd have to fully unpack it first and then compare which doesn't really help with being fast or efficient.

I was hoping to use the previously deployed filesystem's actual files as the seed. I should clarify that I deploy this read-only filesystem, then make a writable btrfs snapshot of it. This leaves the originally deployed filesystem intact on-disk, while allowing efficient small changes online. To my understanding,`casync extract --sync=/mnt/btrfs/read-only-original-fs` would do what I want, just much slower than necessary.

I'm OK with depending on this original read-only directory being unmodified (thus making the created index completely valid). btrfs requires explicitly changing a subvolume to read-write, and even then it updates a UUID that I can verify myself before attempting an update. This logic is used by btrfs's built-in send & receive feature to ensure its seed data is intact without checking every file.

Alternatively, I could store the output of `desync tar` in the filesystem. I hope that the chunk store should be able to reflink most of the actual files on disk, making it only take up dirent space.

**Comment by lights0123** [comment]
Though I imagine for other cases where integrity can't be trusted (like the Steam Deck in the original question if read-write was enabled), files in the already-deployed directory could be verified right before they're copied/reflinked.

**Comment by folbricht** [comment]
What's the primary goal in this use-case? Is it to reflink as many files in new a image/directory as possible to be efficient with disk space? Or is that secondary with performance being more important?

If your images were single large files (which could then be mounted), the `desync extract` command would already do this.  Doing this with `untar` would be very difficult. The main issue is that reflinking happens on files, but chunks in a catar can span file boundaries.

I can think of a couple of ways this may be possible, but it's going to be either quite complex to implement, or slow.

1. One could for example first read the old catar and build an index of which chunk is contained in which file(s) and at what offsets. The while while running untar on the new catar, find out what chunk is producing the data stream for each file, and if that chunk exists in the old image, reflink it from there instead.

2. Or, for a simple solution that's not as good. When writing a file during untar, just see if a file of the same name exists in the old dir, open that old file and then compare the streams. Everything up to the first mismatch could be reflinked. This method is reasonably fast, but doesn't take advantage of the chunking algorithm at all. On the plus-side, it wouldn't even require having the old catar available.

**Comment by lights0123** [comment]
Disk space and bandwidth are my primary concerns: if I update a 4kB config file or 1MB binary on the 10GB filesystem, I would like to only use a few MB of disk space and bandwidth to update it. This means that I would not want to download a full catar, but rather a caidx and fill most of the chunks with stuff already on-disk.

> The main issue is that reflinking happens on files, but chunks in a catar can span file boundaries.

There's also [FICLONERANGE](https://www.man7.org/linux/man-pages/man2/ioctl_ficlonerange.2.html) for block-aligned (always 4KiB?) sharing.

> One could for example first read the old catar and build an index of which chunk is contained in which file(s) and at what offsets. The while while running untar on the new catar, find out what chunk is producing the data stream for each file, and if that chunk exists in the old image, reflink it from there instead.

To preface this, I am unfamiliar with the layout of the catar format—this is based on my assumption that it is similar to tar.

I was imagining a format that stays on the target that says, for example, "chunk ab3681... is bytes 0x0000 to 0xFFFF of /abc" or "chunk bf5311... is the header (in `tar` terminology) of /a, then bytes 0x0000 to 0x1000 (the entirety) of /a, then the header of /b, then bytes 0x0000 to 0x1000 (the entirety) of /b". When `extract`ing the caidx, if chunk bf5311 is requested, files /a and /b could just be reflinked to the destination directory.

(slightly more efficient would be to start with a read-write btrfs snapshot of the seed data and update it because `btrfs snapshot` is much faster then reflinking every file, but I don't need this level of efficiency and I imagine it would be hard to maintain consistency since `extract`/`untar` would have to delete files)

**Comment by folbricht** [comment]
The catar format is somewhat similar to regular tar but it has a concept of directories which have header and footer. But if you just consider files, it should work the same in this use-case.

What you're describing is quite similar to 1) above. It could work. Unfortunately it requires the introduction of a new type of directory-index file. We'd need a command to create it, and one to verify it as well since verifying that the seed directory-index matches the seed dir is likely not feasible during the untar stage itself. Best to have the ability to do that separately.

Once you have the directory-index (we need to find a good name for this file) and want to untar a new stream, you'll need to make the stream decoder aware of what chunk is used to produce what section of the stream. If you look at untar.go today, the are some goroutines who fetch the chunks, and others that assemble the stream from them. These layers are currently separate.

**Comment by polarathene** [comment]
> https://github.com/folbricht/desync/issues/242#issuecomment-1630300175
> 
> A chunk inside an archive can span over multiple files so you couldn't really say a file is changed or not until it's unpacked.

> https://github.com/folbricht/desync/issues/242#issuecomment-2154210935
> 
> During untar while reading the stream, how would you know that the file being unpacked matches one already on disk.
> You'd have to fully unpack it first and then compare which doesn't really help with being fast or efficient.

Does this differ from using `--store` + `--index` with `desync tar`?:

```console
# Initial store + seed index:
$ desync tar --no-time -s store -i existing.caidx /root-fs

# An update (create a new index + update the store):
$ touch /root-fs-b/new-file
$ desync tar --no-time -s store -i update-b.caidx /root-fs-b
```

Because with `desync mtree` those same files enable retrieving this information?

```console
$ desync mtree -s store -i existing.caidx
#mtree v1.0
. type=dir mode=0755 uid=0 gid=0 time=0.        0
file type=file mode=0644 uid=0 gid=0 size=24 time=0.000000000 sha512256digest=97b0fc819edb24745c11422b30476acf214a8459d888fb5dda857ee9bb195a5e

$ desync mtree -s store -i update-b.caidx
#mtree v1.0
. type=dir mode=0755 uid=0 gid=0 time=0.        0
file type=file mode=0644 uid=42 gid=0 size=24 time=0.000000000 sha512256digest=97b0fc819edb24745c11422b30476acf214a8459d888fb5dda857ee9bb195a5e
new-dir/new-file type=file mode=0644 uid=0 gid=0 size=0 time=0.000000000 sha512256digest=c672b8d1ef56ed28ab87c3622c5114069bdd3ad7b8f9737498d0c01ecef0967a
```

- It may not be enough information for some use cases, but the common file attributes and content digest are available from that?
- Could you not leverage that to know that `file` had the `uid` modified and `new-dir/new-file` is new content?

**EDIT:** Having read more and looked into the actual store contents I see now that it's chunked a tar of the directory tree, it doesn't have chunks per file.

I'm not 100% sure what the difference between a chunk store directory + `caidx` is compared to a `catar`... presumably more relevant to network transfer (_full tar file as a single request vs many small chunk requests_) and storage (_as archives accumulate vs smaller shared chunk store_).

---

> https://github.com/folbricht/desync/issues/242#issuecomment-1630520392
> 
> In order to successfully apply that we would need a free space that is at least 2x the full update size.
> That's a substantial tradeoff compared to the `caibx` where we don't need any reserved free space at all.

> Correct, if you can't keep the catar on disk on the target it's going to require some extra time, either for processing, or for transferring.

Was this referring to keeping a `catar` from the original deployment stored locally at the client/destination, with updates syncing smaller deltas of an updated `catar` to minimize bandwidth, at the expense of redundant disk usage?

Then extracting the contents of the `catar` locally via `desync untar` with each update? (_which appears to affect all files, [even if there was no change](https://github.com/folbricht/desync/issues/269)_)

---

As I haven't quite yet grokked `desync`, I assume I'm missing some context but if `desync` can compare two index files to know what files have actually changed, presumably only chunks that the destination/target requires can be fetched remotely?

> Unfortunately it requires the introduction of a new type of directory-index file.

Ah ok... so this isn't about `caidx` with a store, but `catar` files? Or is it a difference between `caidx` vs `caibx`? The [terminology section](https://github.com/folbricht/desync#terminology) mentions `caidx` as an chunk index for catar archives.

I don't know what this looks like but had the understanding that the `caidx` knows what chunks to get from the store. From the discussion in this thread, while the information of the file tree and content digest is available, along with chunks, there is no connection/mapping of files to specific chunks? Instead it only knows what chunks represent the entire opaque archive as a whole? (_similar to a binary blob with `caibx`?_)

**Comment by polarathene** [comment]
Below is looking at the concern a different way. I'm not sure if it's an appropriate alternative, but rather than seeding at the client it considers an approach with the server creating a delta index of changes + chunk refs, and applying that. Perhaps it's more complexity than needed or doesn't fit well with `desync` 😅 

---

If I'm understanding how it `desync` is working... could I not supply the client `caidx`, perform the diff for files with the new updated `caidx` at the server, and from that create a new `catar` or `caidx` to distribute as a delta? (_more similar to `zsync` functionality_)

At the client side you still have the `caidx` to represent the current state from the previous `untar`.
- No separate "directory-index" format is necessary in this case as the delta `caidx` / `catar` would contain all relevant chunks to retrieve?
- Except it wouldn't preserve deletion, so the original before/after `caidx` is still relevant and this wouldn't benefit `catar` (_if the full updated `catar` were still required_) unless that sort of metadata were encoded and supplied separately? (_hence the "directory-index" requirement?_)

Deletions concern aside, one could easily have the checksum of a `caidx` used for a query to the server to generate the delta index and construct a minimal `catar` from a chunk store that could be stored in a cache for future requests to minimize overhead in processing and transfer concerns (_said cache can expire over time so storage is not wasted on the server thanks to the chunk store_).

`desync` would not need to be concerned with that sort of optimization, but it's something that would be possible if you could use `desync` with a delta chunk index / `catar` this way (_a way to generate one on the server side, and a way to apply it on the client side_).

---

## Example

Server side:
```bash
# Probably a naive way to name the delta?
# Example value: 09f830128d6acfa879dbba37f749445debceab40d448469ced274e5d56115d09
DELTA_ID="$(b3sum --no-names <(b3sum --no-names client.caidx update-b.caidx))"

# Create a delta index, with optional catar chunk bundle:
# NOTE: Could technically infer catar/caidx output based on from/to inputs,
#       yet I could imagine a benefit of caidx inputs but catar for output.
desync delta --store tar-store --from client.caidx --to update-b.caidx --catar "${DELTA_ID}.catar" "${DELTA_ID}.caidx"

# Not sure if this makes sense instead of a `--catar` option above,
# you could perhaps generate a `catar` from an index and chunk store?
desync tar --no-time --store tar-store --delta-index "${DELTA_ID}.caidx" "${DELTA_ID}.catar"
```

Client side:
```bash
# Acquire or derive the delta index name in a similar fashion to above:
DELTA_ID=09f830128d6acfa879dbba37f749445debceab40d448469ced274e5d56115d09

# Use the delta index / catar to extract the difference (unable to infer changes like deletions):
desync untar --store tar-store --index client.caidx --with-delta "${DELTA_ID}.caidx" /root-fs
# Or with catar, perhaps:
desync untar --with-delta "${DELTA_ID}.catar" client.catar /root-fs
# Technically original catar not required, could probably just extract contents via regular `tar`?
desync untar --with-delta "${DELTA_ID}.catar" /root-fs
```

- Depending on size of the update, a server side prepared `catar` might be useful, or just multiple chunk requests directly via an index. Just needs to be adapted to accommodate changes like deletions 😅 
- That wouldn't be aware of existing chunks that could be used for seeding, but I'm not sure how viable that is when the chunk store represents a tar archive instead of individual files? (_presumably non-issue with tar format and CDC_)
- Seeding chunks is more relevant when needing to construct the full tree, even when you're only applying a small update? Or for larger files where the delta index and catar generation is less beneficial?

---

If the chunk store is able to share chunks across tars efficiently with `caidx` files, then what's the impact of that with the delta approach?
- It would be able to take the `client.caidx` and apply the delta to get `update-b.caidx` right? (_assuming the delta index was encoded a little differently to include changes such as deletions_)
- Is there any overlap with indexing to a chunk store of the tar vs `catar`? (_I assume a `catar` might have to bundle all chunks for a file, thus be larger than necessary, rather than a tar containing only the subset of store chunks needed_)

**Comment by folbricht** [comment]
Just to align on terminology first.
- Given `dirA` with multiple files/dirs in it
- A `catar` is simply an archive containing data, metadata, and structure of  `dirA`, very similar to a regular tar
- When chunking the `catar` we end up with a number of chunk files and a `caidx`
- a `caidx` is fundamentally the same as a `caibx` (perhaps with minor differences in header flags), the different suffix is there to indicate that the chunked file was an archive
- All that `caibx`/`caidx` contain is information about which chunk goes where in the underlying blob. They do not contain any information about the blob or what may be inside the archive.

Not sure I fully understand your proposal of having a "delta", but a form of this could work if a partial catar/caidx could be produced, containing only files that have changed. For example:

- Given 2 catar files (or caidx+store, doesn't really matter in this example), `dir-v1.catar` and `dir-v2.catar`
- It'd be possible to extract both at the same time, and storing any file that's different into a new `dir-v1-v2-diff`
- Then one could create a new (presumably smaller) `dir-v1-v2-diff.catar` which could be distributed to clients
- The smaller `dir-v1-v2-diff.catar` would then be untar'ed over the existing tree, overwriting files that have changed
- This ignores deletions of course, and it would always work on whole files only and not just contain a diff within the file

I suspect there are other tools that can diff a tree more efficiently though and produce a delta one could chunk, transfer with desync and apply client-side.

**Comment by polarathene** [comment]
Verbose response follows. Lacking time to condense it. We seem to be roughly on the same page 👍 

---

> All that `caibx`/`caidx` contain is information about which chunk goes where in the underlying blob.
> They do not contain any information about the blob or what may be inside the archive.

I was just going by the args for `desync mtree`, where I only provided a store and index and received a file list that the index was representing for a `.catar` archive.

Perhaps the mtree information was implicitly inferred by `desync` from other data?

---

## Delta proposal

> Not sure I fully understand your proposal of having a "delta", but a form of this could work if a partial catar/caidx could be produced, containing only files that have changed

Looks like I didn't reference [this comment](https://github.com/folbricht/desync/issues/269#issuecomment-2345215428) on a related issue for the delta diff example:

```bash
# A diff between the two:
$ diff <(desync mtree -s store -i existing.caidx) <(desync mtree -s store -i update-b.caidx)
3a4
> new-file type=file mode=0644 uid=0 gid=0 size=0 time=0.000000000 sha512256digest=c672b8d1ef56ed28ab87c3622c5114069bdd3ad7b8f9737498d0c01ecef0967a

# Just the unique digests from the diff output:
$ diff <(desync mtree -s store -i existing.caidx) <(desync mtree -s store -i update-b.caidx) \
  | grep -oP 'sha512256digest=\K.+' | sort -u
c672b8d1ef56ed28ab87c3622c5114069bdd3ad7b8f9737498d0c01ecef0967a
```

It used the `desync mtree` command to compare indexes before (v1) vs after (v2) with the same store to determine a diff (_rather naively, but roughly communicates the intent_). Seemed useful since it provides the file path, common file attributes, and a digest of the file digest itself.

The linked example was very basic, it only added a new file, no modification or deletions (_where the diff would need to be computed better_).

After the naive diff example, I later realized that the store only managed chunks that composed the `catar` itself:
- Blocks are only shared within the store if the two individual `catar` for before and after archives have some overlap.
- A store index represents the associated asset (blob or archive) as opaque chunks, no internal knowledge as you've pointed out, the filepaths and their metadata is part of the catar (_which `desync mtree` presumably can piece together from store and index to query_).
- There is no mapping present between those catar chunks and individual files, which is [a separate index that was proposed prior to my comment above](https://github.com/folbricht/desync/issues/242#issuecomment-2156418059).

---

## Delta proposal - Additional context

My delta proposal was focused on however `desync mtree` was implemented - having the shared store with both index files for before/after archives (_or v1 + v2, whatever makes most sense as an example_), you would be able to compute a diff that produces a third archive which is a minimal delta of the v2 change to apply for v1 (_representing any new file additions or modifications, while deletions would require additional context since extraction appears naive via replacement of the full file tree?_)

Since v1 isn't being extracted again, it would assume that the target location represents the state of v1 (_an alternative to seeding from the target, which would no longer provide a benefit?_).

My proposal assumes that if this functionality were available:
- A client could send their current `caidx` from a prior update/deployment, or just a checksum to identify it (_probably simpler and easy to cache the result of this request_).
- The server then takes that context of the clients index and the update it's to receive and can generate a smaller catar / index of only the relevant store chunks. That delta from two indexes only needs to be computed once and can be cached for subsequent clients.
- Client receives delta (_catar, or index with remote store chunks to retrieve_) and applies it. The next time this client updates it would indicate to the server it's current state is the v2 index.
- **NOTE:** This client/server functionality described doesn't need to be handled by `desync`, it's just how the feature might be used.

> * Then one could create a new (presumably smaller) `dir-v1-v2-diff.catar` which could be distributed to clients

From an additive only perspective, the delta should always be smaller than a full v2 archive? The delta may (_relative to v1_) be much larger, or it may only be some small file changes.

Since the delta represents only relevant changes for migrating the extraction target from the v1 index to the v2 index, only those changes are actually written to the extraction target. No need to fully construct a v2 catar to represent the full dataset redundantly?

As proposed, it doesn't make any sense to pursue due to caveats with deletions not being supported.

---

## Alternatives to desync (if an XY problem)

> I suspect there are other tools that can diff a tree more efficiently though and produce a delta one could chunk, transfer with desync and apply client-side.

Perhaps? I assumed that since `desync` manages the chunk store and has index or archive formats related to that, it'd be more suitable for deriving a diff/delta of that same format.

From what I understand separate tooling to create `catar` / `caidx` files for `desync` may be a bit awkward, especially for extraction that it's unclear if `desync` would remain relevant in that scenario?

The proposal I provided may not be aligned with what `desync` is intended for, as I could use say `rsync` / `zync`. Those lack some features like reconstructing reflinks however.

I just figured `desync` could potentially be smarter in this scenario, especially when only a small subset changed between archives, seeding the full tree and modifying every file (_when there is no actual difference between v1 and v2 for the majority of files_) seems unnecessary?

Your earlier proposal would have been less optimal AFAIK if seeding a large tree, which as per my linked issue I assume would still overwrite mtime regardless for all files? Or would extracted content as a seed avoid the redundant update?

---

## Alternative - Chunk individual files + support new tree format (metadata + filepaths mapped to chunks)

A third approach may be chunking and indexing individual files into the store directly, with a way to generate the equivalent of `desync mtree` for file attributes and path within the context of a full tree?

I don't know enough about chunk store management, but if you compress chunks of files as content and then optionally archive that (for perf advantage of sequential I/O over network), instead of archiving contents into a bundle (tar) that is chunked and compressed, would there really be much difference? (_the archive of file chunks in this sense is a subset of the store, extracting it like a tar without `desync` would not work_)

When I first read about `casync`, I initially thought the chunk store was at the file level and archives were just a delivery optimization or convenience for reconstructing files. File-level would be the current blob chunking instead of catar AFAIK, but  with support for a tree of files to chunks mapping/index that avoids the discussed drawbacks of chunking a directory with `catar`.

This approach is perhaps the most optimal way?
- The tree index itself is separate from file content. Similar to `.caidx` but with the additional context needed.
  - File attributes and paths are only relevant to the tree itself. Most attributes could dedup reasonably well with the right layout for storage, or just rely on compression (_zstd can train dictionaries to further enhance that_).
  - Each file path being unique can point to the content digest of a file that it represents, which would have the relevant chunks mapped (_which again is an easy optimization with layout to minimize repetition bloat if that's a concern_).
  - Has all the context to derive a delta at the file chunk level too. If sourcing chunks from a remote store, you could deterministically request a stream of chunks, fixed ranges, or bundles (partial or full).
- File-level chunks means content across v1 and v2 that doesn't change retains the same identical chunks.
  - That's not necessarily the case when the files are archived and reference chunks due to some slight differences / shift. You could have whatever permutations of file trees and attributes without any added overhead in storage. I'm aware of how the chunking process can minimize that, but overall seems file-level without the archive abstraction would be more optimal?
  - Easy verification of integrity at the destination for seeding chunks of a specific file. Separate tree index mapping avoids the indirect chunks relation to individual files issue identified with archives.

Probably the most work to implement though 😓 I'm not sure how viable it is for an external tool to implement such a format that could leverage it with `desync`.

**Comment by folbricht** [comment]
The way `desync mtree` works is that it uses the index to reassemble the catar stream in memory (not all at once) and untars the stream. Rather than writing to the filesystem, it just hashes the data in each file. That's why all it needs is the index and the chunks to produce the mtree output.

I'm not really seeing any good options. I'd rather not introduce a new file-format to describe a file-tree with individually chunked files, though I think it be possible in principle and would provide the seed-capabilities.

If the goal is to avoid writing the content in full during `untar`, that's be doable with an incremental mode where it could only replace files that have changed (whole files), something like `desync untar --incremental <archive> <target>`. It still have to unpack the full stream though (no seeding from existing trees). And it probably couldn't handle deletions.
